core/src/main/kotlin/org/eln2/data/MultiMap.kt

```diff
@@ -0,0 +1,231 @@
+	 */
+	val entries: Iterable<Map.Entry<K, V>>
+
+	/**
+	 * The number of pairings in this MultiMap.
+	 *
+	 * A naive, O(n) implementation might be entries.count(), but you are encouraged to implement a faster one if at all possible. Implementors should document their complexity.
+	 */
+	val entriesSize: Int
+}
+
+/**
+ * An implementor of [MultiMap] using the Kotlin stdlib.
+ *
+ * This class is defined here using a Map<K, Set<V>> idiom which is compatible with the Kotlin stdlib. Other, more performant implementations exist, and can be swapped in here if needed.
+ */
+class SetMapMultiMap<K, V>(iter: Iterator<Pair<K, V>>): MultiMap<K, V> {
+	/**
+	 * Internally implements the MultiMap interface as a Map<K, Set<V>>.
+	 */
+	val map: Map<K, Set<V>> = with(mutableMapOf<K, MutableSet<V>>()) {
+		iter.forEach { (k, v) -> getOrPut(k, { mutableSetOf() }).add(v) }
+		println("SMMM.map: $this")
+		entries.associate { (k, v) -> Pair(k, v.toSet()) }
+	}.toMap()
+
+	constructor(): this(emptyList<Pair<K, V>>().iterator())
+
+	override fun get(k: K): Set<V> = map.get(k) ?: emptySet()
+
+	override val keyMapping: Iterable<Map.Entry<K, Set<V>>> = map.entries
+	/** This is an O(1) implementation of this property. */
+	override val keyMappingSize: Int get() = map.size
+	override val keys: Set<K> get() = map.keys
+	override val valueSets: Collection<Set<V>> get() = map.values
+	override val uniqueValues: Set<V> get() = map.values.fold(emptySet()) { next, acc -> acc.union(next) }
+	override val entries: Iterable<Map.Entry<K, V>> get() = map.entries.flatMap { (k, vs) -> vs.map { MultiMapEntry(k, it) }}
+	/** This implementation is O(keyMappingSize)--linear in the number of keys. */
+	override val entriesSize: Int
+		get() = map.entries.map { (_, vs) -> vs.size }.sum()
+}
+
+/**
+ * A Mutable version of the [MultiMap] interface.
+ */
+interface MutableMultiMap<K, V>: MultiMap<K, V> {
+	/**
+	 * Get the set of values for the key [k].
+	 *
+	 * Note that the return is mutable; one can add or remove entries from this map by mutating it.
+	 */
+	override operator fun get(k: K): MutableSet<V>
+
+	/**
+	 * Add an association from [k] to [v] (in addition to any other associations involving [k] or [v]).
+	 */
+	operator fun set(k: K, v: V) = get(k).add(v)
+
+	/**
+	 * Clear all values associated with the key [k].
+	 */
+	fun clear(k: K)
+
+	/**
+	 * Remove an association between [k] and [v].
+	 *
+	 * If the map no longer contains [k], [clear] is invoked (as it usually performs some implementation-specific cleanup).
+	 */
+	fun remove(k: K, v: V) {
+		val s = get(k)
+		s.remove(v)
+		if (s.isEmpty()) clear(k)
+	}
+}
+
+/**
+ * An implementor of [MutableMultiMap] using the Kotlin stdlib.
+ *
+ * This class is defined here using a MutableMap<K, MutableSet<V>> idiom which is compatible with the Kotlin stdlib. Other, more performant implementations exist, and can be swapped in here if needed.
+ */
+class MutableSetMapMultiMap<K, V>: MutableMultiMap<K, V> {
```

Comment:
If you want to reduce the duplication, I think you can implement SetMapMultiMap in terms of MutableSetMapMultiMap, restricting the interface appropriately. (And wow is that a mouthful.)

Personally, I'm fine with this.

Reply:
It's secretly Java, isn't everything supposed to be a mouthful? :)

I... might take you up on that, actually; it's no issue having the underlying mutable structure, I suppose, if you don't expose it. (This assumes "immutable casts" are cheap, like `MutableSet::toSet`, which, incidentally, for this hypothetical implementation, the `MutableSetMapMultiMap::toSetMapMultiMap()` would be--it returns `this`.)